verilog/tools/preprocessor/verilog_preprocessor.cc

```diff
@@ -42,25 +51,177 @@ static absl::Status StripComments(const SubcommandArgsRange& args,
+  std::string source_view{post_preproc};
+  verilog::VerilogAnalyzer analyzer(source_view, "file1", config);
+  auto analyze_status = analyzer.Analyze();
+  /* const auto& mydata = analyzer.Data().Contents(); */
+  /* outs<<mydata; */
+
+  /* TODO(karimtera): regarding conditionals
+   1) Modify VerilogPreprocess config to have a configuration to generate SV
+   source codes for all possible variants. 2) Then use parser, directly from
+   VerilogAnalyzer or from VerilogParser to have less dependences. 3) Now, we
+   should have multiple trees, we need to merge them as described by Tom in
+   Verible's issue. 4) Finally, travese the tree and output the chosen path
+   based on definitions.
+  */
+  return absl::OkStatus();
+}
+
+static absl::Status MultipleCU(const SubcommandArgsRange& args, std::istream&,
+                               std::ostream& outs, std::ostream&) {
+  if (args.empty()) {
+    return absl::InvalidArgumentError("Missing file arguments.");
+  }
+  for (const char* source_file : args) {
+    outs << source_file << ":\n";
+    auto status = PreprocessSingleFile(source_file, std::cin, outs, std::cerr);
+    if (!status.ok()) return status;
+    outs << '\n';
+  }
+  return absl::OkStatus();
+}
+
+static absl::Status GenerateVariants(const SubcommandArgsRange& args,
+                                     std::istream&, std::ostream& outs,
+                                     std::ostream&) {
+  const int limit_variants = absl::GetFlag(FLAGS_limit_variants);
+  if (args.size() > 1) {
+    return absl::InvalidArgumentError(
+        "ERROR: generate-variants only works on one file.");
+  }
+  const char* source_file = args[0];
+  std::string source_contents;
+  if (auto status = verible::file::GetContents(source_file, &source_contents);
+      !status.ok()) {
+    std::cerr << "ERROR: passed file can't be open\n.";
+    return status;  // check if the the file is not readable or doesn't exist.
+  }
+
+  // Lexing the input SV source code.
+  verilog::VerilogLexer lexer(source_contents);
+  verible::TokenSequence lexed_sequence;
+  for (lexer.DoNextToken(); !lexer.GetLastToken().isEOF();
+       lexer.DoNextToken()) {
+    // For now we will store the syntax tree tokens only, ignoring all the
+    // white-space characters. however that should be stored to output the
+    // source code just like it was.
+    if (verilog::VerilogLexer::KeepSyntaxTreeTokens(lexer.GetLastToken())) {
+      lexed_sequence.push_back(lexer.GetLastToken());
+    }
+  }
+
+  // Control flow tree constructing.
+  verilog::FlowTree control_flow_tree(lexed_sequence);
+  int counter = 0;
+  auto status = control_flow_tree.GenerateVariants(
+      [limit_variants, &counter](const verilog::FlowTree::Variant& variant) {
+        if (counter == limit_variants) return false;
+        counter++;
+        std::cout << "Variant number " << counter << ":\n";
+        for (auto token : variant.sequence) std::cout << token << '\n';
```

Comment:
is this meant to be to `std::cout` ? Looks like `outs` is the stream we should be targeting.
This looks like it prints the raw tokens, separated with newline which might not be as useful. Ideally, we get a re-creation of a processable SystemVerilog file, but with the conditional blocks removed.

Reply:
Yes, this should be `outs`.
I added the support for white-spaces in PR #1376 so that I can recreate the SystemVerilog file after preprocessing.
Can we keep the output for now as tokens until we get there?